sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala

```diff
@@ -245,7 +301,21 @@ class InMemoryCatalog extends ExternalCatalog {
           s"'$db' table '$table':\n$dupSpecsStr")
       }
     }
-    parts.foreach { p => existingParts.put(p.spec, p) }
+
+    val tableDir = new Path(catalog(db).db.locationUri, table)
+    // TODO: we should follow hive to roll back if one partition path failed to create.
+    parts.foreach { p =>
+      if (p.storage.locationUri.isEmpty) {
+        val partitionPath = p.spec.map { case (k, v) => s"$k=$v" }.mkString("/")
```

Comment:
I am not 100% sure if this always works. Here, do we support all the possible inputs for `key` and `value`? Any limit when we create a directory name?